Add workspace-hack crate to unify third party dependency features

metajack opened the pull request:
By unifying the features requested in third party dependencies, this crate
improves caching and should speed up build/test time. Each time cargo is
invoked, it resolves features for its targets, but since the targets change
every invocation potentially, the resolved features for third party crates
might also change, causing those third party crates to be recompiled. This
unifies the features among third party crates so that every feature resolution
is the same for third party crates.

The downside is we have dependencies to libra-workspace-hack from every first
party crate, and we create new, unused dependency edges from some crates to
third party dependencies. Both downsides can be mitigated by using cargo-guppy
with the `--omit-edges-into libra-workspace-hack`.

This solution is modeled on the one in the rust compiler, which is at
https://github.com/rust-lang/rust/tree/master/src/tools/rustc-workspace-hack

sausagee commented on common/workspace-hack/src/lib.rs:
```diff
@@ -0,0 +1,2 @@
+// intentionally left blank
+
```
nit - our internal git check doesn't like blank line at EOF.

sunshowers commented:
Looks good. Could you file an issue for me to add guppy lints to ensure:
* every crate depends on it
* the exact feature set is captured (+ we have a command to generate the cargo.toml)?

davidiw commented on consensus/safety-rules/Cargo.toml:
```diff
@@ -20,6 +20,7 @@ libra-secure-push-metrics = { path = "../../secure/push-metrics", version = "0.1
 libra-secure-storage = { path = "../../secure/storage", version = "0.1.0" }
 libra-temppath = { path = "../../common/temppath", version = "0.1.0" }
 libra-types = { path = "../../types", version = "0.1.0" }
+libra-workspace-hack = { path = "../../common/workspace-hack", version = "0.1.0" }
```
While I love what this will do for build times, it brings in features that might have been explicitly disabled to minimize dependencies. Is there another way to accomplish this via some ci specific run script?

metajack replied:
I'll have to look and see what the delta is.

metajack commented:
I'm happy to delay landing until we can chat more about this. Adding @mimoo here so he can him in too. You bring up a good point about tooling breaking, although I believe dephell now does everything we need and uses guppy as the source of truth about the dependency graph (which we can easily manipulate to accomodate this change).

bors-libra commented:
:umbrella: The latest upstream changes (presumably bb4c7343083264a24c78cb74538093f177778113) made this pull request unmergeable. Please resolve the merge conflicts.

metajack commented:
Rebased and added a lint to ensure it's used everywhere.

Note that I check that this only includes deps already in the TCBs by cross referencing with:
`cargo guppy select --target x86_64-unknown-linux-gnu --kind thirdparty safety-rules executor libra-secure-push-metrics libra-secure-net libra-secure-time libra-transaction-scripts libra-key-manager libra-vault-client libra-secure-storage`

sunshowers commented on devtools/x/src/lint/guppy.rs:
```diff
@@ -146,3 +146,51 @@ impl PackageLinter for CrateNamesPaths {
+    ) -> Result<RunStatus<'l>> {
+        let package_id = ctx.metadata().id();
+        let pkg_graph = ctx.project_ctx().package_graph().expect("can't find package graph");
+        let workspace_hack_id = pkg_graph.packages().find_map(|metadata| {
+            if metadata.name() == "libra-workspace-hack" {
+                return Some(metadata.id());
+            }
+            None
```
You can search through just the workspace packages for a cheaper query. (I have a task filed to add a by-package-name index.)

sunshowers replied:
actually, you can just search through the links of each package if you like as well.

sunshowers replied:
`pkg_graph.dep_links(package.id()).unwrap()` should give you a list of links, and then you can look at `to.name()` to see if it's correct.

metajack replied:
Done.

sunshowers commented on devtools/x/src/lint/guppy.rs:
```diff
@@ -146,3 +146,51 @@ impl PackageLinter for CrateNamesPaths {
+        }
+
+        if let Some(links) = pkg_graph.dep_links(package_id) {
+            let mut has_links = false;
```
Why is `has_links` necessary?

metajack replied:
If there are no deps at all, there's no need to add workspace-hack. So we emit the lint only if it has deps *and* workspace-hack is not one of them.

metajack replied:
I tried for a while to find some version of find/map/filter, etc that would do this, but I was not successful.

metajack commented:
@bors-libra r=sunshowers

bors-libra commented:
:pushpin: Commit ea934d4 has been approved by `sunshowers`

<!-- @bors-libra r=sunshowers ea934d4d9817c88a53d6d5b0743696c91f351a36 -->

bors-libra commented:
:hourglass: Testing commit ea934d4d9817c88a53d6d5b0743696c91f351a36 with merge 532e910f8b6b33c863752a7024119754325b125f...

bors-libra commented:
:broken_heart: Test failed - [checks-circle_commit_workflow](https://circleci.com/workflow-run/309ea5d6-8a04-4cee-8a16-3f7af1552504?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)

metajack commented:
@bors-libra r=sunshowers

bors-libra commented:
:pushpin: Commit 3fee43d has been approved by `sunshowers`

<!-- @bors-libra r=sunshowers 3fee43d93e19c7417350521889ae0f427465d8c1 -->

bors-libra commented:
:hourglass: Testing commit 3fee43d93e19c7417350521889ae0f427465d8c1 with merge 9eff5b5ff48764beac96cd50b924a1661411e469...

bors-libra commented:
:sunny: Test successful - [checks-actions_land_blocking_test](https://github.com/libra/libra/runs/584419769), [checks-circle_commit_workflow](https://circleci.com/workflow-run/9dce3d63-b6c3-421f-80af-39d495c756a8?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
Approved by: sunshowers
Pushing 9eff5b5ff48764beac96cd50b924a1661411e469 to master...
